File: android/app/src/main/java/com/judgemycal/app/domain/SwapEngine.kt
package com.judgemycal.app.domain

/**
 * On-device fallback for the health_shopping capability, ported from the
 * backend's `suggest_swaps`. Neutral, goal-aligned suggestions — nothing is
 * ever labelled "junk", "cheat", or "bad".
 */
object SwapEngine {

    private val SWAPS = mapOf(
        "white rice" to "brown rice or basmati — similar taste, steadier energy",
        "fries" to "roasted potato wedges — same comfort, less oil",
        "beef burger" to "a leaner turkey or grilled-chicken burger, if you fancy it",
        "naan" to "roti — lighter, still great with curry",
        "soda" to "sparkling water with lime",
    )

    fun suggestSwaps(items: List<String>): SwapResult {
        val out = items.mapNotNull { item ->
            val key = item.trim().lowercase()
            SWAPS.entries
                .firstOrNull { (k, _) -> k in key || key in k }
                ?.let { SwapSuggestion(item = item, swap = it.value) }
        }
        return SwapResult(
            swaps = out,
            note = "Suggestions only — your call, no pressure.",
            fromFallback = true,
        )
    }
}

File: webapp/src/domain/swapEngine.ts
import type { SwapResult } from "./models";

/**
 * In-browser fallback for the health_shopping capability, ported from the
 * backend's `suggest_swaps`. Neutral, goal-aligned suggestions — nothing is
 * ever labelled "junk", "cheat", or "bad".
 */

const SWAPS: Record<string, string> = {
  "white rice": "brown rice or basmati — similar taste, steadier energy",
  fries: "roasted potato wedges — same comfort, less oil",
  "beef burger": "a leaner turkey or grilled-chicken burger, if you fancy it",
  naan: "roti — lighter, still great with curry",
  soda: "sparkling water with lime",
};

export function suggestSwaps(items: string[]): SwapResult {
  const swaps = items.flatMap((item) => {
    const key = item.trim().toLowerCase();
    const hit = Object.entries(SWAPS).find(
      ([k]) => key.includes(k) || k.includes(key),
    );
    return hit ? [{ item, swap: hit[1] }] : [];
  });
  return {
    swaps,
    note: "Suggestions only — your call, no pressure.",
    fromFallback: true,
    personaReaction: "",
  };
}
